README.md
# wysibotbs

![](https://media.tenor.com/gIRYjRZoJqgAAAAC/wysi-727.gif)


CHANGELOG.md
# Changelog

## [1.1.0](https://github.com/ChecksumDev/wysibot/compare/v1.0.0...v1.1.0) (2023-11-11)


### Features

* save scores and update logging ([44afaf0](https://github.com/ChecksumDev/wysibot/commit/44afaf0357a9316dec66985eb13045f963add92c))


### Bug Fixes

* correct database structure, use TDResult ([37764cd](https://github.com/ChecksumDev/wysibot/commit/37764cd6db95b55a946ee55467b1f0f3f8cd9307))
* remove `#` from user prefix ([6271a63](https://github.com/ChecksumDev/wysibot/commit/6271a637b40f01de46284cae5dc22451a96cf9e5))

## 1.0.0 (2023-11-08)


### Features

* init ([bd7e024](https://github.com/ChecksumDev/wysibot/commit/bd7e02475fe6fd0ad37e9f28863aa9702965b8e1))
* rewrite entire project ([7426aeb](https://github.com/ChecksumDev/wysibot/commit/7426aebe130393b919b9f01ec2882814eddd4084))


### Bug Fixes

* automatic reconnect to websocket ([08dd140](https://github.com/ChecksumDev/wysibot/commit/08dd1406bc75016d906d87bdccb716a9f12663f1))
* automatic token refreshing and code cleanup ([801bd13](https://github.com/ChecksumDev/wysibot/commit/801bd13f388c8d201cae1631a818ead0613862a4))
* clip delay + async refactor ([4439afd](https://github.com/ChecksumDev/wysibot/commit/4439afd73b1812d0b7d547d693c0176473ed17f9))
* had to do this on a chromebook, oops! ([536ef46](https://github.com/ChecksumDev/wysibot/commit/536ef46ea01d87bac06f143a06150f33bef74734))
* replace old database logic ([b7e8c98](https://github.com/ChecksumDev/wysibot/commit/b7e8c9833ae7613d64ae840805ee7372d7b25223))
* save twitch tokens as `json` in database ([9aebe49](https://github.com/ChecksumDev/wysibot/commit/9aebe4925adaec79a6ceee25d9022511be860c4c))


### Reverts

* ci ([e0982a8](https://github.com/ChecksumDev/wysibot/commit/e0982a8a21c814f73ced0569cdf6702f58945c4a))


src/main.ts
import { Logger, createLogger, format, transports } from 'winston'
import { TwitchService } from './services/TwitchService'
import { TwitterService } from './services/TwitterService'
import { insertScore } from './utils'
import Database from 'bun:sqlite'
import fs from 'fs'


class Client {
    // logger
    private logger: Logger

    // database
    private db: Database

    // services
    private ws: WebSocket | null = null
    private twitchService: TwitchService
    private twitterService: TwitterService

    constructor() {
        this.logger = createLogger({
            level: 'debug',
            format: format.combine(
                format.timestamp({
                    format: 'YYYY-MM-DD HH:mm:ss',
                }),
                format.errors({ stack: true }),
                format.splat(),
                format.json()
            ),
            transports: [
                new transports.File({ filename: 'logs/error.log', level: 'error' }),
                new transports.File({ filename: 'logs/combined.log' }),
            ],
        })

        if (process.env.NODE_ENV !== 'production') {
            this.logger.add(
                new transports.Console({
                    format: format.combine(
                        format.colorize(),
                        format.printf(info => `${info.timestamp} [${info.service || 'main'}::${info.level}]: ${info.message}`)
                    ),
                })
            )
        }

        fs.mkdirSync('data', { recursive: true })

        this.db = new Database('data/db.sqlite')
        this.initDatabase()

        // initialize services
        this.twitchService = new TwitchService(this.logger, this.db)
        this.twitterService = new TwitterService(this.logger, this.db)
    }


    private initDatabase() {
        this.db.exec(`
            CREATE TABLE IF NOT EXISTS twitchAuth (
                id INTEGER PRIMARY KEY,
                tokenData TEXT NOT NULL
            )
        `)

        this.db.exec(`
            CREATE TABLE IF NOT EXISTS twitterAuth (
                id INTEGER PRIMARY KEY,
                tokenData TEXT NOT NULL
            )
        `)

        this.db.exec(`
            CREATE TABLE IF NOT EXISTS scores (
                id TEXT PRIMARY KEY,
                songid TEXT NOT NULL,
                playerid TEXT NOT NULL,
                score INTEGER NOT NULL,
                accuracy REAL NOT NULL,
                timestamp INTEGER NOT NULL
            )
        `)
    }

    private connectWebSocket(): void {
        this.ws = new WebSocket('wss://api.beatleader.xyz/scores')

        this.ws.addEventListener('open', () => {
            this.logger.info('Connected to BeatLeader')
        })

        this.ws.addEventListener('message', async (event: any) => {
            let score = JSON.parse(event.data)
            await this.onScore(score)
        })

        this.ws.addEventListener('close', () => {
            this.logger.info('Disconnected from BeatLeader, reconnecting in 5 seconds...')
            setTimeout(() => this.connectWebSocket(), 5000)
        })
    }

    public async onScore(score: any): Promise<void> {
        const rawAcc = (Math.round(score.accuracy * 10000) / 100).toString()
        const acc = rawAcc.replace('.', '')

        this.logger.debug(`${score.id} by ${score.player.name} (${score.player.id}) - ${rawAcc}% on ${score.leaderboard.song.name} (${score.leaderboard.difficulty.difficultyName})`)
        if (!acc.includes('727')) return

        let player = await fetch(`https://api.beatleader.xyz/player/${score.playerId}?stats=true`).then(res => res.json())

        let twitchSocial = player.socials.find((social: any) => social.service === 'Twitch')
        let twitterSocial = player.socials.find((social: any) => social.service === 'Twitter')

        let username = player.name
        let url = `https://replay.beatleader.xyz/?scoreId=${score.id}`

        if (twitchSocial) {
            let twitchUrl = await this.twitchService.parseScore(twitchSocial, score)
            if (twitchUrl) url = twitchUrl
        }

        if (twitterSocial) {
            username = `@${twitterSocial.link.split("/")[3]}`
        }

        let tweet = `#WYSI ${username} just got ${rawAcc}% on ${score.leaderboard.song.name} (${score.leaderboard.difficulty.difficultyName}) on #BeatSaber! ${url}`
        await this.twitterService.twitterClient.v2.tweet(tweet)

        await this.twitchService.chatClient.say(
            "wysibot",
            `${score.player.name} just scored ${rawAcc}% on ${score.leaderboard.song.name} (${score.leaderboard.difficulty.difficultyName}) ${url}`
        )

        insertScore(this.db, score)
        this.logger.info(`WYSI ${score.player.name} just scored ${rawAcc}% on ${score.leaderboard.song.name} (${score.leaderboard.difficulty.difficultyName}) ${url}`)
    }


    public async run(): Promise<void> {
        this.logger.info('Starting up...')

        // validate environment variables
        this.validateEnvironmentVariables()

        // run services
        await this.twitchService.run()
        await this.twitterService.run()

        this.connectWebSocket()
    }

    private validateEnvironmentVariables(): void {
        this.logger.debug('Validating environment variables...')
        const requiredVariables = [
            // twitch
            'TWITCH_USER_ID',
            'TWITCH_CLIENT_ID',
            'TWITCH_CLIENT_SECRET',
            // twitter
            'TWITTER_CLIENT_ID',
            'TWITTER_CLIENT_SECRET',
        ]

        const missingVariables = requiredVariables.filter(
            (variable) => !process.env[variable]
        )

        if (missingVariables.length > 0) {
            this.logger.error(
                `Missing environment variables: ${missingVariables.join(', ')}`
            )
            process.exit(1)
        }

        this.logger.debug('Environment variables are valid!')
    }
}

const client = new Client()
await client.run()


src/services/TwitchService.ts
import { RefreshingAuthProvider, exchangeCode } from "@twurple/auth"
import { insertOrUpdateToken } from "../utils"
import { ApiClient } from "@twurple/api"
import { ChatClient } from "@twurple/chat"
import { Logger } from "winston"
import Database from "bun:sqlite"
import TDResult from "../TDResult"

export class TwitchService {
    private logger: Logger
    private authProvider: RefreshingAuthProvider
    public apiClient: ApiClient
    public chatClient: ChatClient
    private db: Database

    constructor(logger: Logger, db: Database) {
        this.logger = logger.child({ service: "twitch" })
        this.db = db

        this.authProvider = new RefreshingAuthProvider({
            clientId: process.env.TWITCH_CLIENT_ID!,
            clientSecret: process.env.TWITCH_CLIENT_SECRET!,
        })

        this.authProvider.onRefresh((id: number, newTokenData: any) => {
            insertOrUpdateToken(this.db, id, newTokenData, "twitchAuth")
            this.logger.info(`Refreshed Twitch token for user ${id}`)
        })

        this.authProvider.onRefreshFailure((id: number) => {
            this.logger.error(`Failed to refresh Twitch token for user ${id}`)
        })

        this.apiClient = new ApiClient({ authProvider: this.authProvider })
        this.chatClient = new ChatClient({
            authProvider: this.authProvider,
            channels: ["wysibot"],
        })
    }

    private async setupAuthProvider() {
        const tokenQuery = this.db.prepare("SELECT * FROM twitchAuth WHERE id = ?")
        let result = tokenQuery.get(process.env.TWITCH_USER_ID!) as TDResult | undefined

        if (!result) {
            this.logger.info("No token data found, please authorize with the following URL:")
            this.logger.info(`https://id.twitch.tv/oauth2/authorize?response_type=code&client_id=${process.env.TWITCH_CLIENT_ID!}&redirect_uri=http://localhost:3000&scope=clips:edit+chat:read+chat:edit+whispers:read+whispers:edit+channel:bot+user:bot`)
            this.logger.info("Then paste the code from the redirect URL here:")

            let code = await new Promise<string>((resolve, _) => {
                process.stdin.resume()
                process.stdin.once("data", (data) => {
                    resolve(data.toString().trim())
                })
                1
            })

            const tokenData = await exchangeCode(
                process.env.TWITCH_CLIENT_ID!,
                process.env.TWITCH_CLIENT_SECRET!,
                code,
                "http://localhost:3000"
            )

            this.authProvider.addUser(process.env.TWITCH_USER_ID!, tokenData, ["chat"])
            insertOrUpdateToken(this.db, process.env.TWITCH_USER_ID!, tokenData, "twitchAuth")
        } else {
            this.authProvider.addUser(process.env.TWITCH_USER_ID!, JSON.parse(result.tokenData), ["chat"])
        }
    }

    public async parseScore(social: any, score: any): Promise<string | undefined> {
        let user = await this.apiClient.users.getUserById(social.userId)
        if (!user) return

        let url = `https://twitch.tv/${user.name}`
        let stream = await this.apiClient.streams.getStreamByUserId(user.id)
        if (stream) {
            await this.apiClient.asUser(process.env.TWITCH_USER_ID!, async (apiClient) => {
                try {
                    let clip = await apiClient.clips.createClip({
                        channel: user!.id,
                        createAfterDelay: true,
                    })

                    if (clip) url = `https://clips.twitch.tv/${clip}`
                } catch (e) {
                    this.logger.warn(`Failed to create clip for ${user!.displayName}!`)
                }
            })
        }

        await this.chatClient.join(user.name)
        await this.chatClient.say(
            user.name,
            `! WHEN YOU SEE IT! You just got a ${(Math.round(score.accuracy * 10000) / 100)}% on ${score.leaderboard.song.name} (${score.leaderboard.difficulty.difficultyName}) ${url}`
        )

        return url
    }

    async run(): Promise<void> {
        this.logger.info("Starting up...")
        await this.setupAuthProvider()
        this.chatClient.connect()
    }
}


src/services/TwitterService.ts
import { TwitterApi, TwitterApiReadWrite } from "twitter-api-v2"
import { TwitterApiAutoTokenRefresher } from "@twitter-api-v2/plugin-token-refresher"
import { Logger } from "winston"
import Database from "bun:sqlite"
import { insertOrUpdateToken } from "../utils"
import TDResult from "../TDResult"

export class TwitterService {
    private logger: Logger
    private db: Database
    public twitterClient: TwitterApiReadWrite

    constructor(logger: Logger, db: Database) {
        this.logger = logger.child({ service: "twitter" })
        this.db = db

        this.twitterClient = new TwitterApi({
            clientId: process.env.TWITTER_CLIENT_ID!,
            clientSecret: process.env.TWITTER_CLIENT_SECRET!,
        })
    }

    private createTwitterClient(refreshToken: string): TwitterApiReadWrite {
        return new TwitterApi({
            clientId: process.env.TWITTER_CLIENT_ID!,
            clientSecret: process.env.TWITTER_CLIENT_SECRET!,
        }, {
            plugins: [new TwitterApiAutoTokenRefresher({
                refreshToken: refreshToken,
                refreshCredentials: {
                    clientId: process.env.TWITTER_CLIENT_ID!,
                    clientSecret: process.env.TWITTER_CLIENT_SECRET!,
                },
                onTokenUpdate: (newTokenData) => {
                    insertOrUpdateToken(this.db, 0, newTokenData, "twitterAuth")
                    this.logger.info("Refreshed Twitter token!")
                },
                onTokenRefreshError: (error) => {
                    this.logger.error("Failed to refresh Twitter token!")
                    this.logger.error(error)
                }
            })]
        }).readWrite
    }


    private async initTwitter(): Promise<any> {
        const tokenQuery = this.db.prepare("SELECT * FROM twitterAuth WHERE id = ?")
        let result = tokenQuery.get(0) as TDResult | undefined

        if (!result) {
            let auth = this.twitterClient.generateOAuth2AuthLink("http://localhost:3000", { scope: ["offline.access", "users.read", "tweet.read", "tweet.write"] })

            this.logger.info("No token data found, please authorize with the following URL:")
            this.logger.info(auth.url)
            this.logger.info("Then paste the code from the redirect URL here:")

            let code = await new Promise<string>((resolve, _) => {
                process.stdin.resume()
                process.stdin.once("data", (data) => {
                    resolve(data.toString().trim())
                })
                1
            })

            const { client: _, accessToken, refreshToken } = await this.twitterClient.loginWithOAuth2({
                code: code,
                codeVerifier: auth.codeVerifier,
                redirectUri: "http://localhost:3000"
            })

            let tokenData = { accessToken, refreshToken }
            insertOrUpdateToken(this.db, 0, tokenData, "twitterAuth")
            this.twitterClient = this.createTwitterClient(refreshToken!) // exists because of offline.access scope
        } else {
            this.twitterClient = this.createTwitterClient(JSON.parse(result.tokenData).refreshToken)
        }
    }

    public async run(): Promise<void> {
        await this.initTwitter()
    }
}

src/index.ts
import { TwitterApi, TwitterApiReadWrite } from "twitter-api-v2";
import { TwitterApiAutoTokenRefresher } from "@twitter-api-v2/plugin-token-refresher";
import { RefreshingAuthProvider, exchangeCode } from "@twurple/auth";
import { Database } from "bun:sqlite";
import { ApiClient } from "@twurple/api";
import { ChatClient } from "@twurple/chat";

interface Data {
  value: string;
}

class WYSIBot {
  private twitterClient: TwitterApiReadWrite | undefined;

  // twitch
  private twitchAuthProvider: RefreshingAuthProvider;
  private twitchApi: ApiClient;
  private twitchChat: ChatClient;

  private db: Database;

  constructor() {
    if (!process.env.TWITTER_CLIENT_ID || !process.env.TWITTER_CLIENT_SECRET) {
      throw new Error(
        "Please set TWITTER_CLIENT_ID and TWITTER_CLIENT_SECRET environment variables."
      );
    }

    this.twitchAuthProvider = new RefreshingAuthProvider({
      clientId: process.env.TWITCH_CLIENT_ID!,
      clientSecret: process.env.TWITCH_CLIENT_SECRET!,
    });

    this.twitchAuthProvider.onRefresh(async (_: number, newTokenData: any) => {
      this.updateTokens("twitch", JSON.stringify(newTokenData));
    });

    this.twitchApi = new ApiClient({ authProvider: this.twitchAuthProvider });

    this.twitchChat = new ChatClient({
      authProvider: this.twitchAuthProvider,
      channels: ["wysibot"],
    });

    this.db = new Database("./data/db.sqlite", { create: true });
    this.db.exec(`CREATE TABLE IF NOT EXISTS settings (
            key TEXT PRIMARY KEY,
            value TEXT
        );`);
  }

  connectBeatLeader() {
    let websocket = new WebSocket("wss://api.beatleader.xyz/scores");

    websocket.onopen = () => {
      console.log("Connected to BeatLeader.");
    };

    websocket.onclose = () => {
      console.log("Disconnected from BeatLeader, reconnecting...");
      setTimeout(() => {
        this.connectBeatLeader();
      }, 1000);
    };

    websocket.onmessage = async (event) => {
      let data =
        event.data instanceof Buffer ? event.data.toString() : event.data;
      let score = JSON.parse(data);
      await this.onScore(score);
    };
  }

  async run() {
    console.log("Starting...");
    await this.initTwitch();
    await this.initTwitter();

    this.twitchApi = new ApiClient({ authProvider: this.twitchAuthProvider });
    this.twitchChat.connect();

    this.connectBeatLeader();
  }

  private async onScore(score: any) {
    try {
      const acc = (Math.round(score.accuracy * 10000) / 100).toString();
      const accString = acc.replace(".", "");
      if (!accString.includes("727")) return;

      const response = await fetch(
        `https://api.beatleader.xyz/player/${score.playerId}?stats=true`
      );

      const player = await response.json();

      let url = `https://replay.beatleader.xyz/?scoreId=${score.id}`;

      const twitterSocial = player.socials.find(
        (social: any) => social.service === "Twitter"
      );

      const playername = twitterSocial
        ? `@${twitterSocial.link.split("/")[3]}`
        : player.name;

      const twitchSocial = player.socials.find(
        (social: any) => social.service === "Twitch"
      );

      if (twitchSocial) {
        let twitchUser = await this.twitchApi.users.getUserById(
          twitchSocial.userId
        );
        if (twitchUser?.name) {
          let stream = await this.twitchApi.streams.getStreamByUserName(
            twitchUser.name
          );

          if (stream) {
            url = `https://twitch.tv/${twitchUser.name}`;

            try {
              let clip = await this.twitchApi.clips.createClip({
                channel: twitchSocial.userId,
              });
              if (clip) {
                url = clip;
              }
            } catch (e) {
              console.error(e);
            }
          }

          await this.twitchChat.join(twitchUser.name);
          await this.twitchChat.say(
            twitchUser.name,
            `! WHEN YOU SEE IT! You just got ${acc}% on ${
              score.leaderboard.song.name
            } (${score.leaderboard.difficulty.difficultyName}) ${
              stream ? url : ""
            }`
          );
        }
      }

      let tweet = `#WYSI ${playername} just got ${acc}% on ${score.leaderboard.song.name} (${score.leaderboard.difficulty.difficultyName}) on #BeatSaber! ${url}`;
      await this.twitchChat.say(
        "wysibot",
        `${score.player.name} just got ${acc}% on ${score.leaderboard.song.name} (${score.leaderboard.difficulty.difficultyName}) ${url}`
      );

      if (this.twitterClient) await this.twitterClient.v2.tweet(tweet);

      console.log(tweet);
    } catch (error) {
      console.error("Error in onScore:", error);
    }
  }

  private updateTokens(type: string, data: string) {
    let existing = this.db
      .prepare(`SELECT value FROM settings WHERE key = '${type}'`)
      .all() as Data[];
    if (existing.length === 0) {
      this.db
        .prepare(
          `INSERT INTO settings (key, value) VALUES ('${type}', '${data}')`
        )
        .run();
    } else {
      this.db
        .prepare(`UPDATE settings SET value = '${data}' WHERE key = '${type}'`)
        .run();
    }
  }

  private getTokens(type: string): any | null {
    let existing = this.db
      .prepare(`SELECT value FROM settings WHERE key = '${type}'`)
      .all() as Data[];
    if (existing.length === 0) {
      return null;
    }

    return JSON.parse(existing[0].value);
  }

  private async initTwitch(): Promise<any> {
    let tokenData = this.getTokens("twitch");

    if (!tokenData) {
      console.log("No tokens found, starting OAuth2 flow...");
      return await this.getFirstTwitchRefreshToken();
    }

    this.twitchAuthProvider.addUser(process.env.TWITCH_USER_ID!, tokenData, [
      "chat",
    ]);
  }

  private async getFirstTwitchRefreshToken() {
    console.log(
      `https://id.twitch.tv/oauth2/authorize?response_type=code&client_id=${process.env.TWITCH_CLIENT_ID}&redirect_uri=http://localhost:3000&scope=analytics:read:extensions+user:edit+user:read:email+clips:edit+bits:read+analytics:read:games+user:edit:broadcast+user:read:broadcast+chat:read+chat:edit+channel:moderate+channel:read:subscriptions+whispers:read+whispers:edit+moderation:read+channel:read:redemptions+channel:edit:commercial+channel:read:hype_train+channel:read:stream_key+channel:manage:extensions+channel:manage:broadcast+user:edit:follows+channel:manage:redemptions+channel:read:editors+channel:manage:videos+user:read:blocked_users+user:manage:blocked_users+user:read:subscriptions+user:read:follows+channel:manage:polls+channel:manage:predictions+channel:read:polls+channel:read:predictions+moderator:manage:automod+channel:manage:schedule+channel:read:goals+moderator:read:automod_settings+moderator:manage:automod_settings+moderator:manage:banned_users+moderator:read:blocked_terms+moderator:manage:blocked_terms+moderator:read:chat_settings+moderator:manage:chat_settings+channel:manage:raids+moderator:manage:announcements+moderator:manage:chat_messages+user:manage:chat_color+channel:manage:moderators+channel:read:vips+channel:manage:vips+user:manage:whispers+channel:read:charity+moderator:read:chatters+moderator:read:shield_mode+moderator:manage:shield_mode+moderator:read:shoutouts+moderator:manage:shoutouts+moderator:read:followers+channel:read:guest_star+channel:manage:guest_star+moderator:read:guest_star+moderator:manage:guest_star`
    );
    console.log("Please enter the code from the URL:");
    let code = await new Promise<string>((resolve, _) => {
      process.stdin.resume();
      process.stdin.once("data", (data) => {
        resolve(data.toString().trim());
      });
      1;
    });

    let tokenData = await exchangeCode(
      process.env.TWITCH_CLIENT_ID!,
      process.env.TWITCH_CLIENT_SECRET!,
      code,
      "http://localhost:3000"
    );
    this.updateTokens("twitch", JSON.stringify(tokenData));

    return await this.initTwitch();
  }

  private async initTwitter(): Promise<any> {
    let tokenData = this.getTokens("twitter");

    console.log("Checking for existing tokens...");
    if (!tokenData) {
      console.log("No tokens found, starting OAuth2 flow...");
      return await this.getFirstTwitterRefreshToken();
    } else {
      const autoRefresherPlugin = new TwitterApiAutoTokenRefresher({
        refreshToken: this.getTokens("twitter")?.refreshToken,
        refreshCredentials: {
          clientId: process.env.TWITTER_CLIENT_ID!,
          clientSecret: process.env.TWITTER_CLIENT_SECRET!,
        },
        onTokenUpdate: (token) => {
          console.log("Refreshing Twitter tokens...");
          let tokenData = {
            accessToken: token.accessToken,
            refreshToken: token.refreshToken,
          };
          this.updateTokens("twitter", JSON.stringify(tokenData));
        },
        onTokenRefreshError: (error) => {
          console.error("Refresh error", error);
        },
      });

      const twitter = new TwitterApi(
        {
          clientId: process.env.TWITTER_CLIENT_ID!,
          clientSecret: process.env.TWITTER_CLIENT_SECRET!,
        },
        {
          plugins: [autoRefresherPlugin],
        }
      );

      this.twitterClient = twitter.readWrite;
    }
  }

  private async getFirstTwitterRefreshToken() {
    const temporary_twitter = new TwitterApi({
      clientId: process.env.TWITTER_CLIENT_ID!,
      clientSecret: process.env.TWITTER_CLIENT_SECRET!,
    });

    let oauth = temporary_twitter.generateOAuth2AuthLink(
      "http://localhost:3000",
      { scope: ["offline.access", "users.read", "tweet.read", "tweet.write"] }
    );
    console.log(oauth.url);
    console.log("Please enter the code from the URL:");

    let code = await new Promise<string>((resolve, _) => {
      process.stdin.resume();
      process.stdin.once("data", (data) => {
        resolve(data.toString().trim());
      });
    });

    const {
      client: _,
      accessToken,
      refreshToken,
    } = await temporary_twitter.loginWithOAuth2({
      code: code,
      redirectUri: "http://localhost:3000",
      codeVerifier: oauth.codeVerifier,
    });
    let tokenData = { accessToken: accessToken, refreshToken: refreshToken };
    this.updateTokens("twitter", JSON.stringify(tokenData));

    return await this.initTwitter();
  }
}

const bot = new WYSIBot();
bot.run();
